File: 1814-jump-game-vi/jump-game-vi.cpp
class Solution {
public:
    int maxResult(vector<int>& nums, int k) {
        deque<pair<int,int>> dq; // {score, index}
        dq.push_back({nums[0], 0});

        for(int i = 1; i < nums.size(); i++){
            // remove out-of-range elements
            while(dq.front().second < i - k)
                dq.pop_front();

            int cur = dq.front().first + nums[i];

            // maintain decreasing order of score
            while(!dq.empty() && dq.back().first <= cur)
                dq.pop_back();

            dq.push_back({cur, i});
        }
        return dq.back().first;
    }
};

File: 1814-jump-game-vi/jump-game-vi.java
// class Solution {
//     public int maxResult(int[] nums, int k) {
//         int n = nums.length, a= 0, b=0;
//         int[] deq = new int[n];
//         deq[0] = n-1;
//         for(int i = n-2; i>=0;i--){
//             if(deq[a] - i > k){
//                 a++;
//             }
//             nums[i] += nums[deq[a]];
//             while(b >= a && nums[deq[b]] <= nums[i]){
//                 b--;
//             }
//             deq[++b]=i;
//         }
//         return nums[0];
//     }
// }
/* class Solution {
    public int maxResult(int[] nums, int k) {
        int n = nums.length;
        int max = nums[0];
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b)-> b[1]-a[1]);
        pq.offer(new int[]{0, nums[0]});
        for(int i = 1; i < n ; i++) {
            while(i - pq.peek()[0] > k) {
                pq.poll();
            } 
            int top[] = pq.peek();
            max = nums[i] + top[1];
            pq.offer(new int[]{i, max});
        }
        return max;
    }
} */



class Solution {
    public int maxResult(int[] nums, int k) {
        int max = nums[0];
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b)-> b[1]-a[1]);
        pq.offer(new int[]{0 , nums[0]});
        for(int i = 1; i < nums.length; i++) {
            while(i - pq.peek()[0] > k) {
                pq.poll();
            }
            int [] top = pq.peek();
            max = nums[i] + top[1];
            pq.offer(new int[]{i, max});
        }
        return max;
    }
}